File: backend/others/Areas.py
import math

# 1 grado son 111,32 km
# 1/111320 es 1 metro
meter = 1/111320
SCALE = meter*10


class Areas(object):

	# Dados dos vértices devuelve una lista con cuatro vértices que conforman un rectángulo
	# El rectángulo devuelto se caracteriza por que uno de sus pares de lados es igual de ´
	# largo que la distancia AB y el otro de sus pares de lados es de tamaño 2*SCALE
	def buildRectangle(self, A, B):
		vector = [B[0]-A[0], B[1]-A[1]]
		if vector[0] > vector[1]:
			vector = [vector[0]/vector[0], vector[1]/vector[0]]
		else:
			vector = [vector[0]/vector[1], vector[1]/vector[1]]
		perpendicular1 = [-vector[1]*SCALE, vector[0]*SCALE]
		perpendicular2 = [vector[1]*SCALE, -vector[0]*SCALE]

		A1 = [A[0]+perpendicular1[0], A[1]+perpendicular1[1]]
		A2 = [A[0]+perpendicular2[0], A[1]+perpendicular2[1]]
		B1 = [B[0]+perpendicular1[0], B[1]+perpendicular1[1]]
		B2 = [B[0]+perpendicular2[0], B[1]+perpendicular2[1]]
		# El orden importa, de este modo se devuelven vértices contiguos
		return [A1, A2, B2, B1]

	# Devuelve el área de un triángulo dados sus vértices en coordenadas
	def calculateTriangleArea(self, A, B, C):
		return 1/2 * abs(A[0]*(B[1]-C[1]) + B[0]*(C[1]-A[1]) + C[0]*(A[1]-B[1]))

	# Devuelve el área de un rectángulo dados sus vértices en coordenadas en orden contiguo
	def calculateRectangleArea(self, A, B, C, D):
		base = math.sqrt((B[0]-A[0])**2 + (B[1]-A[1])**2)
		height = math.sqrt((C[0]-B[0])**2 + (C[1]-B[1])**2)
		# print(base);
		# print(height);
		return base*height

# a=Areas()
# print(a.buildRectangle([0,0],[2,3]))
#print(a.calcularAreaTriangulo([0,0], [2,0], [2,3]))
#print(a.calculateRectangleArea([6.2, 5.0], [3.8, -1.0], [11.2, 3.0], [8.8, -3.0]))

File: backend/administration_app/spring-boot-quartz-demo-master/src/main/java/com/emotionalmap/service/Areas.java
package com.emotionalmap.service;

import java.util.ArrayList;

import com.emotionalmap.entity.Position;

public class Areas {

	// 1 grado son 111,32 km
	// 1/111320 es 1 metro
	private double meter = 1 / 111320;
	private double SCALE = meter * 10;

	public ArrayList<Position> buildRectangle(Position A, Position B) {
		ArrayList<Double> vector = new ArrayList<Double>();
		ArrayList<Double> vectorAux = new ArrayList<Double>();
		vector.add(B.getLng() - A.getLng());
		vector.add(B.getLat() - A.getLat());
		if (vector.get(0) > vector.get(1)) {
			vectorAux.add(vector.get(0) / vector.get(0));
			vectorAux.add(vector.get(1) / vector.get(0));
		} else {
			vectorAux.add(vector.get(0) / vector.get(1));
			vectorAux.add(vector.get(1) / vector.get(1));
		}

		ArrayList<Double> perpendicular1 = new ArrayList<Double>();
		ArrayList<Double> perpendicular2 = new ArrayList<Double>();

		perpendicular1.add(-vector.get(1) * SCALE);
		perpendicular1.add(vector.get(0) * SCALE);
		perpendicular2.add(vector.get(1) * SCALE);
		perpendicular2.add(-vector.get(0) * SCALE);

		Position A1 = new Position(A.getLng() + perpendicular1.get(0), A.getLat() + perpendicular1.get(1));
		Position A2 = new Position(A.getLng() + perpendicular2.get(0), A.getLat() + perpendicular2.get(1));
		Position B1 = new Position(B.getLng() + perpendicular1.get(0), B.getLat() + perpendicular1.get(1));
		Position B2 = new Position(B.getLng() + perpendicular2.get(0), B.getLat() + perpendicular2.get(1));

		// El orden importa, de este modo se devuelven vértices contiguos
		ArrayList<Position> res = new ArrayList<Position>();
		res.add(A1);
		res.add(A2);
		res.add(B2);
		res.add(B1);
		return res;
	}

	public double calculateRectangleArea(Position A, Position B, Position C, Position D) {
		double base = Math.sqrt(Math.pow((B.getLng() - A.getLng()), 2) + Math.pow((B.getLat() - A.getLat()), 2));
		double height = Math.sqrt(Math.pow((C.getLng() - B.getLng()), 2) + Math.pow((C.getLat() - B.getLat()), 2));
		return base * height;
	}

	public double calculateTriangleArea(Position A, Position B, Position C) {
		return 1 / 2 * Math.abs(A.getLng() * (B.getLat() - C.getLat()) + B.getLng() * (C.getLat() - A.getLat())
				+ C.getLng() * (A.getLat() - B.getLat()));
	}

}